README.md

```diff
@@ -1,10 +1,260 @@
+
+In order have fine grained control over pytest command, execute:
+
+	make shell
+
+This will launch an interactive environment inside the container that has all
+the dependencies installed. The developer may now launch the `pytest` command
+the way it's necessary, debug the environment and temporarily add/change
+the dependencies.
+
+### Makefile
+Makefile provides an easy way to start the testing environment without the need
+to worry about the correct docker commands. It's core concept is `adminrouter-devkit`
+container which is contains all the dependencies that are needed to run
+Adminrouter, and inside which all tests-related commands are run.
+
+It exposes a couple of targets:
+* `make clean` - remove all containers created by the test harness. It does not
+  remove the images themselves though as the layer cache may be useful later
+  on and the user may remove them themselves.
+* `make devkit` - creates the `adminrouter-devkit` container. By default other
+   targets execute this step automatically if devkit container image does not
+   exist yet.
+* `make update-devkit` - updates `adminrouter-devkit`. Should be run every time
+   the Dockerfile or its dependencies change.
+* `make tests` - launch all the tests. Worth noting is the fact that mccabe
+  complexity of the code is also verified and an error is raised if it's equal
+  or above 10.
+* `make shell` - launch an interactive shell within the devkit container. Should
+  be used in case when fine grained control of the tests is necessary or
+  during debugging.
+* `make flake8` - launch flake8 which will check all the tests and test-harness
+  files by default.
+
+### Docker container
+As mentioned earlier, all the commands are executed inside the `adminrouter-devkit`
+container`. It follows the same build process for Nginx that happens during
+DC/OS build with the execption of setting the  `--with-debug` flag. It also
+contains some basic debugging tools, py.test related dependencies and basic
+files that help py.test mimic the DC/OS environment. Their location is then
+exported via environment variables to the py.test code, so changing their location
+can be done by edition only the Dockerfile.
+
+Due to some issues with overlayfs on older kernels, the container links
+/tmp directory to /dev/shm.
+
+### Repository flavours
+Adminrouter reposity can come in two variants/flavours - Enterprise and Opensource.
+Tests determine it basing on the directory structure - if tests contain
+`test-harness/tests/open/` directory then the repo is treated as Opensource one,
+and in case when `test-harness/tests/ee/` directory is present - as the
+enterprise one. Pytest fixture `repo_is_ee` takes care of it and is pulled in
+as dependencies by all the fixtures that need this information.
+
+### Service startup ordering and cleanup
+Mocking out DC/OS is a complex task, and a huge part in it have pytest fixture
+feature. It takes care of proper ordering of mocks and subprocess startups, and
```

Comment:
`Mocking out DC/OS is a complex task, and a huge part in it have pytest fixtures. They take care of proper ordering of mocks and subprocess start/stop and cleanup and the cleanup ordering itself.`

Does it sound better ?